### PR TITLE
Scrolling Refactor

### PR DESCRIPTION
**Pull request by Hixie**

### CODE REVIEW

**Comment by Hixie** [comment]
This is the widget side, not including things like scroll bars, lazy block, and app bars, which I'll submit separately.

The tests here are still inadequate, but I wanted to submit the PR for review while I worked on more tests.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 1]
```diff
@@ -2681,6 +2681,16 @@ class WidgetToRenderBoxAdapter extends LeafRenderObjectWidget {
   }
 }
 
+class SliverToBoxAdapter extends SingleChildRenderObjectWidget {
+  SliverToBoxAdapter({
+    Key key,
+    Widget child,
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/basic.dart]: `@required` ?

File: packages/flutter/lib/src/widgets/overscroll_indicator.dart [hunk 2]
```diff
@@ -0,0 +1,745 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async' show Timer;
+import 'dart:math' as math;
+
+import 'package:flutter/foundation.dart' show required, Listenable, ChangeNotifier;
+import 'package:flutter/physics.dart' show nearEqual, Tolerance;
+import 'package:flutter/rendering.dart' show applyGrowthDirectionToAxisDirection, AxisDirection, GrowthDirection;
+import 'package:flutter/scheduler.dart' show Ticker;
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/overscroll_indicator.dart]: We don't usually trim down these sorts of imports.

File: packages/flutter/lib/src/widgets/overscroll_indicator.dart [hunk 3]
```diff
@@ -0,0 +1,745 @@
...
+import 'package:flutter/widgets.dart';
+
+import 'package:flutter/material.dart' show Theme; // DELETE WHEN REMOVING LEGACY SCROLLING CODE
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/overscroll_indicator.dart]: Oh man

File: packages/flutter/lib/src/widgets/overscroll_indicator.dart [hunk 4]
```diff
@@ -0,0 +1,745 @@
...
+  @override
+  void initState() {
+    super.initState();
+    _leadingController = new _GlowController(vsync: this, color: config.color, axis: config.axis);
+    _trailingController = new _GlowController(vsync: this, color: config.color, axis: config.axis);
+  }
+
+  @override
+  void didUpdateConfig(GlowingOverscrollIndicator oldConfig) {
+    if (oldConfig.color != config.color) {
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/overscroll_indicator.dart]: Check axis?

File: packages/flutter/lib/src/widgets/overscroll_indicator.dart [hunk 5]
```diff
@@ -0,0 +1,745 @@
...
+    _glowController = new AnimationController(vsync: vsync)
+      ..addStatusListener(_changePhase);
+    Animation<double> decelerator = new CurvedAnimation(
+      parent: _glowController,
+      curve: Curves.decelerate,
+    )..addListener(notifyListeners);
+    _glowOpacity = _glowOpacityTween.animate(decelerator);
+    _glowSize = _glowSizeTween.animate(decelerator);
+    _displacementTicker = vsync.createTicker(_tickDisplacement);
+    _color = color;
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/overscroll_indicator.dart]: Why not move this to the initializer list along with _axis.

File: packages/flutter/lib/src/widgets/overscroll_indicator.dart [hunk 6]
```diff
@@ -0,0 +1,745 @@
...
+  }
+
+  // animation of the main axis direction
+  _GlowState _state = _GlowState.idle;
+  AnimationController _glowController;
+  Timer _pullRecedeTimer;
+
+  // animation values
+  Tween<double> _glowOpacityTween = new Tween<double>(begin: 0.0, end: 0.0);
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/overscroll_indicator.dart]: final?

File: packages/flutter/lib/src/widgets/overscroll_indicator.dart [hunk 7]
```diff
@@ -0,0 +1,745 @@
...
+  Animation<double> _glowOpacity;
+  Tween<double> _glowSizeTween = new Tween<double>(begin: 0.0, end: 0.0);
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/overscroll_indicator.dart]: final?

File: packages/flutter/lib/src/widgets/overscroll_indicator.dart [hunk 8]
```diff
@@ -0,0 +1,745 @@
...
+
+  /// The controller for the overscroll glow on the side with positive scroll offsets.
+  ///
+  /// For a vertical downwards viewport, this is the bottom side.
+  final _GlowController trailingController;
+
+  /// The direction of the viewport.
+  final AxisDirection axisDirection;
+
+  static const double PI_OVER_2 = math.PI / 2.0;
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/overscroll_indicator.dart]: PI_OVER_2 -> probably shouldn't be in SHOUTY_CASE

File: packages/flutter/lib/src/widgets/scroll_absolute.dart [hunk 9]
```diff
@@ -0,0 +1,678 @@
...
+  /// The platform whose scroll physics should be implemented.
+  ///
+  /// Defaults to the current platform.
+  TargetPlatform getPlatform(BuildContext context) => defaultTargetPlatform;
+
+  /// The color to use for the glow effect whet [platform] indicates a platform
+  /// that uses a [GlowingOverscrollIndicator].
+  ///
+  /// Defaults to white.
+  Color getAccentColor(BuildContext context) => const Color(0xFFFFFFFF);
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/scroll_absolute.dart]: AccentColor is a material concept...  Maybe a more specific name?  This is used for the glow, right?  getThemeColor ?

File: packages/flutter/lib/src/widgets/scroll_notification.dart [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+  ScrollNotification2({
+    Scrollable2State scrollable,
+  }) : axisDirection = scrollable.config.axisDirection,
+       metrics = scrollable.position.getMetrics(),
+       context = scrollable.context;
+
+  /// The direction that positive scroll offsets indicate.
+  final AxisDirection axisDirection;
+
+  Axis get axis {
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/scroll_notification.dart]: We have this function in a bunch of places.  Maybe factor out into a static function and have all these call through to it?

**Comment by abarth** [comment]
LGTM

File: packages/flutter/lib/src/widgets/basic.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/basic.dart]: It's a render object widget, and the render object can handle not having a child.

File: packages/flutter/lib/src/widgets/overscroll_indicator.dart (hunk 5 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/overscroll_indicator.dart]: why not indeed

File: packages/flutter/lib/src/widgets/overscroll_indicator.dart (hunk 8 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/overscroll_indicator.dart]: I was being consistent with math.PI. Changed to `piOver2` though.

File: packages/flutter/lib/src/widgets/scroll_absolute.dart (hunk 9 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_absolute.dart]: getGlowColor it is.

File: packages/flutter/lib/src/widgets/scroll_notification.dart (hunk 10 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_notification.dart]: Done (axisDirectionToAxis)

Comment (abarth) [inline packages/flutter/lib/src/widgets/scroll_notification.dart]: Are we consistent about fooToBar versus getBarFromFoo?  localToGlobal is certainly an example of the former.

Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_notification.dart]: The only cases of getBarFromFoo that I could find were privates.

File: packages/flutter/test/rendering/mock_canvas.dart [hunk 11]
```diff
@@ -3,16 +3,54 @@
 // found in the LICENSE file.
 
 import 'package:flutter/rendering.dart';
+import 'package:test/test.dart';
 
 class MockCanvas implements Canvas {
   final List<Invocation> invocations = <Invocation>[];
```
Comment (abarth) [inline packages/flutter/test/rendering/mock_canvas.dart]: Can you add a TODO about making this private.

File: packages/flutter/test/rendering/mock_canvas.dart [hunk 12]
```diff
@@ -3,16 +3,54 @@
...
 
+  int _saveCount = 0;
+
+  @override
+  int getSaveCount() => _saveCount;
+
+  @override
+  void save() {
+    _saveCount += 1;
+    invocations.add(new _MethodCall(#save));
```
Comment (abarth) [inline packages/flutter/test/rendering/mock_canvas.dart]: I would have done this with super.save

File: packages/flutter/test/rendering/mock_canvas.dart [hunk 13]
```diff
@@ -3,16 +3,54 @@
...
+  }
+
+  @override
+  void restore() {
+    _saveCount -= 1;
+    assert(_saveCount >= 0);
+    invocations.add(new _MethodCall(#restore));
```
Comment (abarth) [inline packages/flutter/test/rendering/mock_canvas.dart]: ditto

**Comment by Hixie** [comment]
Once Travis goes green I'm landing this.

**Pull request merged**